File: scripts/continue_import.py
import os
import pandas as pd
from supabase import create_client, Client
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Initialize Supabase client
SUPABASE_URL: str = os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_ANON_KEY: str = os.environ.get("NEXT_PUBLIC_SUPABASE_ANON_KEY")

if not SUPABASE_URL or not SUPABASE_ANON_KEY:
    raise ValueError("Please set NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY in .env.local")

supabase: Client = create_client(SUPABASE_URL, SUPABASE_ANON_KEY)

# Load the renamed CSV
print("Loading CSV...")
df = pd.read_csv("data/dpwh_transparency_data_renamed.csv")

print(f"Total rows in CSV: {len(df)}")

# Get all existing contract IDs from the database
print("Fetching existing contract IDs from database...")
existing_contract_ids = set()

# Fetch in batches of 1000
start = 0
batch_size = 1000

while True:
    response = supabase.table('dpwh_projects').select('contract_id').range(start, start + batch_size - 1).execute()
    if not response.data:
        break
    for row in response.data:
        existing_contract_ids.add(row['contract_id'])
    start += batch_size
    print(f"Fetched {len(existing_contract_ids)} existing contract IDs...")

print(f"Total existing contract IDs: {len(existing_contract_ids)}")

# Filter out rows that are already in the database
print("Filtering CSV...")
df_filtered = df[~df['contract_id'].isin(existing_contract_ids)]

print(f"Rows to import: {len(df_filtered)}")

if len(df_filtered) == 0:
    print("All rows are already imported!")
    exit()

# Convert NaN to None
df_filtered = df_filtered.where(pd.notnull(df_filtered), None)

# Convert boolean columns
df_filtered['is_live'] = df_filtered['is_live'].astype(bool)
df_filtered['has_satellite_image'] = df_filtered['has_satellite_image'].astype(bool)

# Convert date columns
df_filtered['start_date'] = pd.to_datetime(df_filtered['start_date'], errors='coerce').dt.date
df_filtered['completion_date'] = pd.to_datetime(df_filtered['completion_date'], errors='coerce').dt.date

# Convert NaT to None
df_filtered['start_date'] = df_filtered['start_date'].where(pd.notnull(df_filtered['start_date']), None)
df_filtered['completion_date'] = df_filtered['completion_date'].where(pd.notnull(df_filtered['completion_date']), None)

# Import in batches of 1000 rows
batch_size = 1000
total_rows = len(df_filtered)
print(f"Importing {total_rows} rows...")

for i in range(0, total_rows, batch_size):
    batch = df_filtered.iloc[i:i + batch_size]
    batch_dict = batch.to_dict('records')
    
    try:
        response = supabase.table('dpwh_projects').upsert(batch_dict, on_conflict='contract_id').execute()
        print(f"Imported rows {i + 1} - {min(i + batch_size, total_rows)} of {total_rows}")
    except Exception as e:
        print(f"Error importing batch {i}: {e}")
        # Try one row at a time for this batch
        print("Trying one row at a time...")
        for j, row in enumerate(batch_dict):
            try:
                supabase.table('dpwh_projects').upsert(row, on_conflict='contract_id').execute()
                print(f"Successfully imported row {i + j + 1}")
            except Exception as e2:
                print(f"Error with row {i + j + 1}: {e2}")
                print(f"Row data: {row}")

print("Import complete!")

File: scripts/continue-import.js
require('dotenv').config({ path: '.env.local' });
const fs = require('fs');
const csv = require('csv-parser');
const { createClient } = require('@supabase/supabase-js');

// Initialize Supabase client
const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_ANON_KEY = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_ANON_KEY) {
  throw new Error('Please set NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY in .env.local');
}

const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

async function continueImport() {
  // Step 1: Read all rows from CSV
  console.log('Reading CSV...');
  const csvRows = [];
  await new Promise((resolve, reject) => {
    fs.createReadStream('data/dpwh_transparency_data_renamed.csv')
      .pipe(csv())
      .on('data', (row) => csvRows.push(row))
      .on('end', resolve)
      .on('error', reject);
  });

  console.log(`Total rows in CSV: ${csvRows.length}`);

  // Step 2: Fetch all existing contract_ids from database
  console.log('Fetching existing contract IDs...');
  const existingContractIds = new Set();
  let start = 0;
  const batchSize = 1000;

  while (true) {
    const { data, error } = await supabase
      .from('dpwh_projects')
      .select('contract_id')
      .range(start, start + batchSize - 1);

    if (error) throw error;
    if (!data || data.length === 0) break;

    data.forEach(row => existingContractIds.add(row.contract_id));
    start += batchSize;
    console.log(`Fetched ${existingContractIds.size} existing contract IDs...`);
  }

  console.log(`Total existing contract IDs: ${existingContractIds.size}`);

  // Step 3: Filter out rows that already exist
  const rowsToImport = csvRows.filter(row => !existingContractIds.has(row.contract_id));
  console.log(`Rows to import: ${rowsToImport.length}`);

  if (rowsToImport.length === 0) {
    console.log('All rows are already imported!');
    return;
  }

  // Step 4: Import remaining rows
  console.log('Starting import...');
  const importBatchSize = 500; // Smaller batches to avoid timeouts

  for (let i = 0; i < rowsToImport.length; i += importBatchSize) {
    const batch = rowsToImport.slice(i, i + importBatchSize);
    
    // Convert types
    const convertedBatch = batch.map(row => ({
      ...row,
      budget: row.budget ? parseFloat(row.budget) : null,
      amount_paid: row.amount_paid ? parseInt(row.amount_paid, 10) : null,
      progress: row.progress ? parseFloat(row.progress) : null,
      is_live: row.is_live === 'True' || row.is_live === 'true' || row.is_live === '1',
      has_satellite_image: row.has_satellite_image === 'True' || row.has_satellite_image === 'true' || row.has_satellite_image === '1',
      latitude: row.latitude ? parseFloat(row.latitude) : null,
      longitude: row.longitude ? parseFloat(row.longitude) : null,
      report_count: row.report_count ? parseInt(row.report_count, 10) : null,
      start_date: row.start_date || null,
      completion_date: row.completion_date || null
    }));

    try {
      const { error } = await supabase
        .from('dpwh_projects')
        .upsert(convertedBatch, { onConflict: 'contract_id' });

      if (error) throw error;
      console.log(`Imported rows ${i + 1} - ${Math.min(i + importBatchSize, rowsToImport.length)} of ${rowsToImport.length}`);
    } catch (error) {
      console.error(`Error importing batch ${i}:`, error.message);
      // Try one by one
      console.log('Trying one row at a time...');
      for (let j = 0; j < convertedBatch.length; j++) {
        try {
          const { error: rowError } = await supabase
            .from('dpwh_projects')
            .upsert(convertedBatch[j], { onConflict: 'contract_id' });
          if (rowError) throw rowError;
          console.log(`Successfully imported row ${i + j + 1}`);
        } catch (rowError) {
          console.error(`Error with row ${i + j + 1}:`, rowError.message);
          console.error('Row data:', convertedBatch[j]);
        }
      }
    }
  }

  console.log('Import complete!');
}

continueImport().catch(console.error);
